### PR TITLE
feat: Apply MetalLB configuration to remote cluster

### PR DESCRIPTION
**Pull request by dlipovetsky**
**What problem does this PR solve?**:
- Generates MetalLB configuration from ServiceLoadBalancer Configuration API (added in #778)
- Waits for MetalLB helm chart to be successfully deployed (using waiter added in #777)
- Applies MetalLB configuration to remote cluster, also using a wait.

To apply MetalLB configuration, the MetalLB CRD webhooks must be ready. This is not guaranteed to happen once the MetalLB helm chart is successfully deployed.

Because the MetalLB configuration is applied in a non-blocking lifecycle hook, the topology controller retries the hook immediately after it returns; we cannot ask for a delay. Because the topology controller backs off its requests exponentially, this often results in the hook requests being delayed for 3+ minutes before succeeding. 

To mitigate this exponential backoff, the hook retries internally for to up to 20 seconds, within the 30 seconds recommended by the Cluster API.

<details>
<summary>Related log excerpt</summary>

```
    I0628 19:39:31.051929       1 handler.go:132] "Deploying ServiceLoadBalancer provider MetalLB" cluster="default/dlipovetsky"
    I0628 19:39:31.051954       1 handler.go:82] "Applying MetalLB installation" cluster="default/dlipovetsky"
    I0628 19:39:31.100706       1 cm.go:82] "Fetching HelmChart info for \"metallb\" from configmap default/default-helm-addons-config" cluster="default/dlipovetsky"
    E0628 19:39:41.031497       1 handler.go:140] "failed to deploy ServiceLoadBalancer provider MetalLB" err="context canceled: last apply error: failed to apply MetalLB configuration IPAddressPool metallb-system/metallb: server-side apply failed: Patch \"https://172.18.0.2:6443/apis/metallb.io/v1beta1/namespaces/metallb-system/ipaddresspools/metallb?fieldManager=cluster-api-runtime-extensions-nutanix&fieldValidation=Strict&timeout=10s\": context canceled" cluster="default/dlipovetsky"
    ...
    E0628 19:43:49.496027       1 handler.go:140] "failed to deploy ServiceLoadBalancer provider MetalLB" err="context canceled: last apply error: failed to apply MetalLB configuration IPAddressPool metallb-system/metallb: server-side apply failed: Internal error occurred: failed calling webhook \"ipaddresspoolvalidationwebhook.metallb.io\": failed to call webhook: Post \"https://metallb-webhook-service.metallb-system.svc:443/validate-metallb-io-v1beta1-ipaddresspool?timeout=10s\": dial tcp 10.130.84.48:443: connect: connection refused" cluster="default/dlipovetsky"
    I0628 19:43:51.785907       1 handler.go:132] "Deploying ServiceLoadBalancer provider MetalLB" cluster="default/dlipovetsky"
    I0628 19:43:51.785923       1 handler.go:82] "Applying MetalLB installation" cluster="default/dlipovetsky"
    I0628 19:43:51.790521       1 cm.go:82] "Fetching HelmChart info for \"metallb\" from configmap default/default-helm-addons-config" cluster="default/dlipovetsky"
```

</details>

**Which issue(s) this PR fixes**:
Fixes #

**How Has This Been Tested?**:

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/handlers/generic/lifecycle/serviceloadbalancer/metallb/handler.go [hunk 1]
```diff
@@ -152,5 +158,77 @@ func (n *MetalLB) Apply(
...
+	})
+	if err != nil {
+		return fmt.Errorf("failed to generate MetalLB configuration: %w", err)
+	}
+
+	var applyErr error
+	if waitErr := kwait.PollUntilContextTimeout(
+		ctx,
+		5*time.Second,
+		10*time.Second,
```
Comment (dkoshkin) [inline pkg/handlers/generic/lifecycle/serviceloadbalancer/metallb/handler.go:200]: Just checking that this is the intended timeout? During testing we were seeing that it backed off pretty aggressively and CAPI didn't call the hook for a few mins.

**Review by dkoshkin — APPROVED** [review]
Thanks for the stacked PRs, makes it very easy to review.

File: pkg/handlers/generic/lifecycle/serviceloadbalancer/metallb/handler.go [hunk 2]
```diff
@@ -152,5 +158,77 @@ func (n *MetalLB) Apply(
...
+			Check: func(_ context.Context, obj *caaphv1.HelmChartProxy) (bool, error) {
+				for _, c := range obj.GetConditions() {
+					if c.Type == caaphv1.HelmReleaseProxiesReadyCondition && c.Status == corev1.ConditionTrue {
+						return true, nil
+					}
+				}
+				return false, nil
+			},
+			Interval: 5 * time.Second,
+			Timeout:  10 * time.Second,
```
Comment (dkoshkin) [inline pkg/handlers/generic/lifecycle/serviceloadbalancer/metallb/handler.go:175]: In the PR description you mentioned 20 seconds.

Comment (dlipovetsky) [reply pkg/handlers/generic/lifecycle/serviceloadbalancer/metallb/handler.go:175]: Good observation.

We want the handler to return within 30 seconds. We should let the context determine that; the handler caller should cancel the context in 30 seconds. Each wait can be increased to 30 seconds; they will be cancelled as soon as the handler context is cancelled. I'll make this change.

Comment (dlipovetsky) [reply pkg/handlers/generic/lifecycle/serviceloadbalancer/metallb/handler.go:175]: The context originates here: https://github.com/kubernetes-sigs/cluster-api/blob/c05b21b5f8710e4171975b1fe57fca68933fdd6f/exp/runtime/server/server.go#L315

It comes from the request, and the godoc says:

> // For incoming server requests, the context is canceled when the
> // client's connection closes, the request is canceled (with HTTP/2),
> // or when the ServeHTTP method returns.

CAPI says:

> Runtime Extension developers can decide how long the Cluster API Runtime should wait for a Runtime Extension to respond before treating the call as a failure (max is 30s) by returning the timeout during discovery. 
> -- https://cluster-api.sigs.k8s.io/tasks/experimental-features/runtime-sdk/implement-extensions.html#timeouts

I guess the client cancels the connection in 30s, triggering the context cancellation.

Comment (dkoshkin) [reply pkg/handlers/generic/lifecycle/serviceloadbalancer/metallb/handler.go]: Thanks for explaining!

**Review by jimmidyson — APPROVED** [review]

**Pull request merged**